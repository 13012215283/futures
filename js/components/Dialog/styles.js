import { StyleSheet } from 'react-native';

import { colors, sizes } from './../themes';
import { alpha } from '../../utils/color';
import { px2dp } from '../../utils/px2dp';

const headerTextColor = '#212121';
const textColorAnd = '#757575';

const style = StyleSheet.create({
  mask: {
    backgroundColor: alpha(colors.black, 0.4),
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  dialogContainer: {
    backgroundColor: colors.white,
    marginLeft: px2dp(86),
    marginRight: px2dp(85),
    minWidth: px2dp(576),
    borderRadius: px2dp(6),
  },
  textContainer: {
    marginLeft: px2dp(48),
    marginTop: px2dp(48),
    marginRight: px2dp(53),
  },
  header: {
    fontSize: sizes.f4,
    color: headerTextColor,
    marginBottom: px2dp(48),
  },
  contentBox: {
    marginBottom: px2dp(32),
  },
  content: {
    fontSize: sizes.f2,
    color: textColorAnd,
    lineHeight: px2dp(48),
    marginBottom: px2dp(32),
  },
  activeContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    marginBottom: px2dp(16),
    height: px2dp(96),
    alignItems: 'center',
  },
  touchable: {
    borderColor: colors[1104],
    alignItems: 'center',
    justifyContent: 'center',
    width: px2dp(160),
    height: px2dp(72),
    marginRight: px2dp(32),
  },
  activeText: {
    flex: 1,
    fontSize: sizes.f3,
    textAlign: 'right',
    textAlignVertical: 'center',
    color: colors[1001],
  },
  text0: {
    color: colors[1102],
  },
});

export default style;
