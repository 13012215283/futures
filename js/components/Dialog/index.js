import React, { Component } from 'react';
import { Modal, Text, View, TouchableOpacity } from 'react-native';
import PropTypes, { oneOf, oneOfType } from 'prop-types';
import styles from './styles';

class Dialog extends Component {
  constructor(props) {
    super(props);
    this.renderButton = this.renderButton.bind(this);
    this.renderContent = this.renderContent.bind(this);
  }

  renderContent() {
    const { content } = this.props;

    if (typeof content === 'function') {
      return <View style={styles.contentBox}>{content()}</View>;
    }

    return content ? <Text style={styles.content}>{content}</Text> : null;
  }

  renderButton() {
    const { button } = this.props;

    // 最多支持传入两个按钮
    return button.slice(0, 2).map((v, k, array) => {
      let textStyle = [styles.activeText];
      if (array.length === 2)
        textStyle = [styles.activeText, styles[`text${k}`]];

      return (
        <TouchableOpacity
          key={v.name || ''}
          style={styles.touchable}
          onPress={() => v.callback()}
        >
          <Text style={[textStyle, v.style]}>{v.name}</Text>
        </TouchableOpacity>
      );
    });
  }

  render() {
    const { header, content, ...props } = this.props;
    return (
      <View>
        <Modal
          onRequestClose={this.props.onRequestClose}
          transparent
          {...props}
        >
          <View style={styles.mask}>
            <View style={styles.dialogContainer}>
              <View style={styles.textContainer}>
                <Text style={styles.header}>{header}</Text>
                {this.renderContent()}
              </View>
              <View style={styles.activeContainer}>{this.renderButton()}</View>
            </View>
          </View>
        </Modal>
      </View>
    );
  }
}

Dialog.propTypes = {
  header: PropTypes.string,
  content: oneOfType([PropTypes.string, PropTypes.func]),
  /* eslint-disable react/forbid-prop-types */
  button: PropTypes.array.isRequired,
  visible: PropTypes.bool,
  animationType: oneOf(['none', 'slide', 'fade']),
  onRequestClose: PropTypes.func,
};

Dialog.defaultProps = {
  header: '提示',
  visible: false,
  animationType: 'fade',
  onRequestClose() {},
  content: '',
};
export default Dialog;
