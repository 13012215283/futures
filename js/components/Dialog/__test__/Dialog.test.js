import React from 'react';
import { shallow } from 'enzyme';
import TestRender from 'react-test-renderer';
import { View, Modal, TouchableOpacity, Alert } from 'react-native';
import Dialog from '../index';

describe('弹窗组件', () => {
  it('能够正确渲染', () => {
    const tree = TestRender.create(
      <Dialog
        content="内容"
        header="标题"
        button={[{ name: '按钮', callback() {} }]}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('组件包括5层View,一个Moal', () => {
    const tree = shallow(
      <Dialog
        content="内容"
        header="标题"
        button={[{ name: '按钮', callback() {} }]}
      />
    );

    expect(tree.find(View).length).toBe(5);
    expect(tree.find(Modal).length).toBe(1);
  });

  it('最多接受传入两个按钮', () => {
    const tree = shallow(
      <Dialog
        content="内容"
        header="标题"
        button={[
          {
            name: '确认',
            callback() {
              Alert.alert('确认');
            },
          },
          { name: '取消', callback() {} },
          { name: '按钮三', callback() {} },
        ]}
      />
    );

    expect(tree.find(TouchableOpacity).length).toBe(2);
  });
});
