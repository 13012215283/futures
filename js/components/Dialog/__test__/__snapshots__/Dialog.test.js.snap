// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`弹窗组件 能够正确渲染 1`] = `
<View>
  <Modal
    animationType="fade"
    button={
      Array [
        Object {
          "callback": [Function],
          "name": "按钮",
        },
      ]
    }
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#00000066",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#ffffff",
            "borderRadius": 6,
            "marginLeft": 86,
            "marginRight": 85,
            "minWidth": 576,
          }
        }
      >
        <View
          style={
            Object {
              "marginLeft": 48,
              "marginRight": 53,
              "marginTop": 48,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#212121",
                "fontSize": 40,
                "marginBottom": 48,
              }
            }
          >
            标题
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#757575",
                "fontSize": 28,
                "lineHeight": 48,
                "marginBottom": 32,
              }
            }
          >
            内容
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 96,
              "justifyContent": "flex-end",
              "marginBottom": 16,
            }
          }
        >
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            collapsable={undefined}
            hasTVPreferredFocus={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "borderColor": "#e0e0e0",
                "height": 72,
                "justifyContent": "center",
                "marginRight": 32,
                "opacity": 1,
                "width": 160,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Array [
                    Object {
                      "color": "#e83b3b",
                      "flex": 1,
                      "fontSize": 32,
                      "textAlign": "right",
                      "textAlignVertical": "center",
                    },
                  ],
                  undefined,
                ]
              }
            >
              按钮
            </Text>
          </View>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
